src/operator/contrib/ctc_loss-inl.h

```diff
@@ -402,19 +395,18 @@ class CTCLossOp : public Operator {
                             label_lengths->data(),
                             data_lengths->data(),
                             costs.dptr_,
-                            grad_desc_,
-                            grad.dptr_,
+                            ctx.is_train?grad_desc_:NULL,
+                            ctx.is_train?grad.dptr_:NULL,
                             ctc_algo,
                             ctc_desc_,
                             work_space.dptr_,
                             workspace_bytes));
-  }
-  inline virtual void cudnn_backward_extra(mshadow::Stream<xpu>* s,
-                                           mshadow::Tensor<xpu, 3, real_t> data_grad,
-                                           mshadow::Tensor<xpu, 1, real_t> output_grad,
-                                           mshadow::Tensor<xpu, 3, real_t> data_grad_computed) {
-    mxnet_op::SoftmaxGrad<mshadow::op::mul, mxnet_op::softmax_bwd>(s,
-        output_grad.dptr_, data_grad_computed.dptr_, data_grad.dptr_, data_grad.shape_, 2);
+
+    if (ctx.is_train) {
```

Comment:
why test for is_train?

Reply:
This is to see whether gradient calculation is needed, similar to the warp ctc case where if `ctx.is_train` is false then only inference is performed.

Reply:
We now allow for calculating gradient for is_train=False